tidy3d/components/heat/simulation.py

```diff
@@ -0,0 +1,815 @@
+from .grid import HeatGridType
+from .viz import HEAT_BC_COLOR_TEMPERATURE, HEAT_BC_COLOR_FLUX, HEAT_BC_COLOR_CONVECTION
+from .viz import plot_params_heat_bc, plot_params_heat_source, HEAT_SOURCE_CMAP
+
+from ..base_sim.simulation import AbstractSimulation
+from ..base import cached_property
+from ..types import Ax, Shapely, TYPE_TAG_STR, ScalarSymmetry, Bound
+from ..viz import add_ax_if_none, equal_aspect, PlotParams
+from ..structure import Structure
+from ..geometry.base import Box, GeometryGroup
+from ..geometry.primitives import Sphere, Cylinder
+from ..geometry.polyslab import PolySlab
+from ..scene import Scene
+
+from ..bc_placement import StructureBoundary, StructureStructureInterface
+from ..bc_placement import StructureSimulationBoundary, SimulationBoundary
+from ..bc_placement import MediumMediumInterface
+
+from ...exceptions import SetupError
+from ...constants import inf, VOLUMETRIC_HEAT_RATE
+
+HEAT_BACK_STRUCTURE_STR = "<<<HEAT_BACKGROUND_STRUCTURE>>>"
+
+HeatSingleGeometryType = (Box, Cylinder, Sphere, PolySlab)
+
+
+class HeatSimulation(AbstractSimulation):
+    """Contains all information about heat simulation.
+
+    Example
+    -------
+    >>> from tidy3d import Medium, SolidSpec, FluidSpec, UniformUnstructuredGrid, TemperatureMonitor
+    >>> heat_sim = HeatSimulation(
+    ...     size=(3.0, 3.0, 3.0),
+    ...     structures=[
+    ...         Structure(
+    ...             geometry=Box(size=(1, 1, 1), center=(0, 0, 0)),
+    ...             medium=Medium(
+    ...                 permittivity=2.0, heat_spec=SolidSpec(
+    ...                     conductivity=1,
+    ...                     capacity=1,
+    ...                 )
+    ...             ),
+    ...             name="box",
+    ...         ),
+    ...     ],
+    ...     medium=Medium(permittivity=3.0, heat_spec=FluidSpec()),
+    ...     grid_spec=UniformUnstructuredGrid(dl=0.1),
+    ...     sources=[UniformHeatSource(rate=1, structures=["box"])],
+    ...     boundary_spec=[
+    ...         HeatBoundarySpec(
+    ...             placement=StructureBoundary(structure="box"),
+    ...             condition=TemperatureBC(temperature=500),
+    ...         )
+    ...     ],
+    ...     monitors=[TemperatureMonitor(size=(1, 2, 3), name="sample")],
+    ... )
+    """
+
+    boundary_spec: Tuple[HeatBoundarySpec, ...] = pd.Field(
+        (),
+        title="Boundary Condition Specifications",
+        description="List of boundary condition specifications.",
+    )
+
+    sources: Tuple[HeatSourceType, ...] = pd.Field(
+        (),
+        title="Heat Sources",
+        description="List of heat sources.",
+    )
+
+    monitors: Tuple[HeatMonitorType, ...] = pd.Field(
+        (),
+        title="Monitors",
+        description="Monitors in the simulation.",
+    )
+
+    grid_spec: HeatGridType = pd.Field(
+        title="Grid Specification",
+        description="Grid specification for heat simulation.",
+        discriminator=TYPE_TAG_STR,
```

Comment:
My understanding was that we should use discriminator whenever variable type is Union with more than 1 element, so that pydantic doesn't try to initialize provided argument with every element of that Union. Do you know if something changed during pydantic 2 transition?

Reply:
I was under the impression that something changed with the discriminator recently, but I think I might be mistaken. let's leave it in.